src/main/kotlin/solve/scene/controller/SceneController.kt

```diff
@@ -49,42 +90,98 @@ class SceneController : Controller() {
             scrollY(value)
         }
 
+    init {
+        val sceneFacade = SceneFacade(this)
+        ServiceLocator.registerService(sceneFacade)
+
+        addGridSettingsBindings()
+    }
+
     fun setScene(newScene: Scene, keepSettings: Boolean) {
-        columnsCount = calculateColumnsCount(newScene)
         sceneProperty.value = newScene
-        if (!keepSettings || scaleProperty.value < minScale) {
-            scaleProperty.value = minScale
+
+        if (!keepSettings) {
+            reinitializeSettings(newScene)
+        } else {
+            recalculateScale()
         }
     }
 
-    fun zoomIn(mousePosition: DoublePoint) = zoom(min(scaleProperty.value * scaleFactor, MaxScale), mousePosition)
+    fun zoomIn(mousePosition: DoublePoint) = zoom(min(scale * ScaleFactor, maxScale), mousePosition)
+
+    fun zoomOut(mousePosition: DoublePoint) = zoom(max(scale / ScaleFactor, min(minScale, scale)), mousePosition)
```

Comment:
You should take minScale given by columns count into account too to avoid this situation:
![image](https://user-images.githubusercontent.com/55746901/232111981-8d3d3fe5-d749-4105-92c5-f91a9576ec4a.png)